src/H5T.c

```diff
@@ -5245,67 +5271,245 @@ H5T__path_find_real(const H5T_t *src, const H5T_t *dst, const char *name, H5T_co
             HGOTO_ERROR(H5E_DATATYPE, H5E_CANTINIT, NULL, "unable to copy datatype for conversion path");
     } /* end if */
     else
-        path = table;
+        path = matched_path;
+
+    /* Initialize the path if it's a new path */
+    if (new_path && H5T__path_find_init_new_path(path, src, dst, conv, &tmp_ctx) < 0)
+        HGOTO_ERROR(H5E_DATATYPE, H5E_CANTINIT, NULL, "unable to initialize new conversion path");
+
+    /* Fail if the path still doesn't have a conversion function at this point */
+    if (!path->conv.u.app_func)
+        HGOTO_ERROR(H5E_DATATYPE, H5E_CANTINIT, NULL, "no appropriate function for conversion path");
+
+    /* Check if paths were inserted into the table through a recursive call
+     * and re-compute the correct location for this path if so. - QAK, 1/26/02
+     */
+    if (old_npaths != H5T_g.npaths)
+        matched_path = H5T__bsearch_path_table(src, dst, &last_cmp, &path_idx);
+
+    /* Replace an existing table entry or add a new entry */
+    if (matched_path && new_path) {
+        herr_t status;
+        int    nprint = 0;
+
+        assert(matched_path == H5T_g.path[path_idx]);
+        H5T__print_stats(matched_path, &nprint);
+
+        /* Free the old conversion path table entry */
```

Comment:
This block of code, to free a conversion path entry, is replicated at least 4 times in this file (search for H5T_CONV_FREE) and would be good to move to a helper routine also.